test/e2e/storage/testsuites/provisioning.go

```diff
@@ -0,0 +1,370 @@
+	limitations under the License.
+*/
+
+package testsuites
+
+import (
+	"fmt"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	"k8s.io/api/core/v1"
+	storage "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/drivers"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+// StorageClassTest represents parameters to be used by provisioning tests
+type StorageClassTest struct {
+	Name               string
+	CloudProviders     []string
+	Provisioner        string
+	Parameters         map[string]string
+	DelayBinding       bool
+	ClaimSize          string
+	ExpectedSize       string
+	PvCheck            func(volume *v1.PersistentVolume) error
+	NodeName           string
+	SkipWriteReadCheck bool
+	VolumeMode         *v1.PersistentVolumeMode
+}
+
+type provisioningTestSuite struct {
+	tsInfo TestSuiteInfo
+}
+
+var _ TestSuite = &provisioningTestSuite{}
+
+// InitProvisioningTestSuite returns provisioningTestSuite that implements TestSuite interface
+func InitProvisioningTestSuite() TestSuite {
+	return &provisioningTestSuite{
+		tsInfo: TestSuiteInfo{
+			name: "provisioning",
+			testPatterns: []testpatterns.TestPattern{
+				testpatterns.DefaultFsDynamicPV,
+			},
+		},
+	}
+}
+
+func (p *provisioningTestSuite) getTestSuiteInfo() TestSuiteInfo {
+	return p.tsInfo
+}
+
+func (p *provisioningTestSuite) skipUnsupportedTest(pattern testpatterns.TestPattern, driver drivers.TestDriver) {
+}
+
+func createProvisioningTestInput(pattern testpatterns.TestPattern, resource provisioningTestResource) provisioningTestInput {
+	driver := resource.driver
+	dInfo := driver.GetDriverInfo()
+	f := dInfo.Framework
+
+	input := provisioningTestInput{
+		testCase: StorageClassTest{
+			ClaimSize:    resource.claimSize,
+			ExpectedSize: resource.claimSize,
+		},
+		cs:    f.ClientSet,
+		pvc:   resource.pvc,
+		sc:    resource.sc,
+		dInfo: dInfo,
+	}
+
+	if dInfo.Config.ClientNodeName != "" {
+		input.testCase.NodeName = dInfo.Config.ClientNodeName
+	}
+
+	return input
+}
+
+func (p *provisioningTestSuite) execTest(driver drivers.TestDriver, pattern testpatterns.TestPattern) {
+	Context(getTestNameStr(p, pattern), func() {
+		var (
+			resource     provisioningTestResource
+			input        provisioningTestInput
+			needsCleanup bool
+		)
+
+		BeforeEach(func() {
+			needsCleanup = false
+			// Skip unsupported tests to avoid unnecessary resource initialization
+			skipUnsupportedTest(p, driver, pattern)
+			needsCleanup = true
+
+			// Setup test resource for driver and testpattern
+			resource = provisioningTestResource{}
+			resource.setupResource(driver, pattern)
+
+			// Create test input
+			input = createProvisioningTestInput(pattern, resource)
```

Comment:
whats the difference between setting up a resource and creating the test input. Are they actually the same thing? I don't like how complicated this all is. We're wrapping structs in other structs then converting them to other structs through a combination of parameters and method recievers. If there's any way we could simplify all of this I would be all for it

Reply:
Agree, I'm not sure if we really need an intermediate `provisioningTestResource`.  I think we can combine them all into one setup method.

Reply:
Thank you for your comment.

> whats the difference between setting up a resource and creating the test input.

"setting up a resource" is to create resources that will be used in testsuites and "creating the test input" is to put all parameters necessary to the testsuits into a single struct.

 My intention is to make ```setupResource``` sharable between testsuites, as we defined ```genericSetupResource```. Most of the testsuites will just create either (a) volumeSource or (b) object definition for creating volumeSource. So, I think that how they create the test resource is common, but which parameters in resource are required as input are different between the testsuites. (In addition, test input will require parameters other than test resource, such as driver parameters and pattern parameters.) As a result, if we combine ```setupResource``` and ```createTestInput```, we would need to implement the same logic for creating resource to all testsuites.

Reply:
can createTestInput for this test suite just call `genericSetupResource` though instead of having a seperate `setupResource` to do it? We don't have to do a huge logic change IIUC